lambdaclass/mirra_backend
# AFK Rewards

Adds AFK rewards.

In Champions of Mirra we have this mechanic called AFK rewards, where beating levels will grant you a different afk rate for currencies. These currencies will be accumulated and whenever you send a `claim` request, you will get whatever you had accumulated since last time you claimed (max. 12 hours).

# Known issues 

We should create new afk reward rates if we give increment_afk_reward_rate a currency that's not owned. Issue: https://github.com/lambdaclass/champions_of_mirra/issues/223

Exploitable bug when accumulating time and advancing levels before claim. Issue: https://github.com/lambdaclass/champions_of_mirra/issues/198

# To test

Create user, get campaigns, get levels, get currencies
```
{:ok, user} = Champions.Users.register("User")
{:ok, [campaign | _]} = Champions.Campaigns.get_campaigns
[level1, level2, level3 | _] = campaign.levels |> Enum.sort_by(& &1.level_number)
gold = GameBackend.Users.Currencies.get_currency_by_name!("Gold")
gems = GameBackend.Users.Currencies.get_currency_by_name!("Gems")
```

Check you have rates and they are `0.0`
```
GameBackend.Rewards.get_afk_reward_rate(user.id, gold.id)
GameBackend.Rewards.get_afk_reward_rate(user.id, gems.id)
```

Beat level 1 (gives only currency)
```
Champions.Battle.fight_level(user.id, level1.id)
```

Check out your gold (should be 110 from the sample currencies in user register + first level beaten)
```
GameBackend.Users.Currencies.get_amount_of_currency(user.id, gold.id)
```

Check your rates have now grown.
```
GameBackend.Rewards.get_afk_reward_rate(user.id, gold.id)
GameBackend.Rewards.get_afk_reward_rate(user.id, gems.id)
```

Wait a few moments and check how much you have accumulated to claim
```
Champions.Users.get_afk_rewards(user.id)
```

Claim it!
```
Champions.Users.claim_afk_rewards(user.id)
```

Check they have been consumed
```
Champions.Users.get_afk_rewards(user.id)
```

And that you have them in your UserCurrencies now!
```
GameBackend.Users.Currencies.get_amount_of_currency(user.id, gold.id)
```
